mobile/fula.go

```diff
@@ -48,66 +64,184 @@ type DirEntry struct {
+	ls    ipld.LinkSystem
+	gx    graphsync.GraphExchange
+	ds    datastore.Batching
 }
 
 func NewFula(repoPath string) (*Fula, error) {
-	err := logging.SetLogLevelRegex(".*", "DEBUG")
-
-	if err != nil {
-		// TODO: don't panic!
-		panic("logger failed")
-	}
 	ctx := context.Background()
-	node, err := create(ctx, repoPath)
+	host, err := create(ctx, repoPath)
 	if err != nil {
 		return nil, err
 	}
-	f := &Fula{Node: *node, ctx: ctx}
+	f := &Fula{h: host, ctx: ctx}
+
+	f.ds = dssync.MutexWrap(datastore.NewMapDatastore())
+
+	f.ls = cidlink.DefaultLinkSystem()
+	f.ls.StorageReadOpener = f.blockReadOpener
+	f.ls.StorageWriteOpener = f.blockWriteOpener
+
+	f.startGraphExchange(ctx)
+
 	return f, nil
 }
 
-func (f *Fula) AddBox(boxAddr string) error {
-	peerAddr, err := peer.AddrInfoFromString(boxAddr)
-	if err != nil {
+func (f *Fula) startGraphExchange(ctx context.Context) {
+
+	gsn := gsnet.NewFromLibp2pHost(f.h)
+	f.gx = gs.New(ctx, gsn, f.ls)
+	f.gx.RegisterIncomingRequestHook(
+		func(p peer.ID, r graphsync.RequestData, ha graphsync.IncomingRequestHookActions) {
+			// TODO check the peer is allowed and request is valid
+			ha.ValidateRequest()
+		})
+}
+
+func (f *Fula) blockWriteOpener(ctx ipld.LinkContext) (io.Writer, ipld.BlockWriteCommitter, error) {
+	buf := bytes.NewBuffer(nil)
+	return buf, func(l ipld.Link) error {
+		return f.ds.Put(ctx.Ctx, toDatastoreKey(l), buf.Bytes())
+	}, nil
+}
+
+func (f *Fula) blockReadOpener(ctx ipld.LinkContext, l ipld.Link) (io.Reader, error) {
+	val, err := f.ds.Get(ctx.Ctx, toDatastoreKey(l))
+	switch err {
+	case nil:
+		return bytes.NewBuffer(val), nil
+	case datastore.ErrNotFound:
+		return nil, format.ErrNotFound{Cid: l.(cidlink.Link).Cid}
+	default:
+		return nil, err
+	}
+}
+
+var lp = cidlink.LinkPrototype{
+	Prefix: cid.Prefix{
+		Version:  1,
+		Codec:    uint64(multicodec.DagCbor),
+		MhType:   uint64(multicodec.Sha2_256),
+		MhLength: -1,
+	},
+}
+
+func (f *Fula) Store(ctx context.Context, n ipld.Node) (ipld.Link, error) {
+	return f.ls.Store(ipld.LinkContext{Ctx: ctx}, lp, n)
+}
+
+func (f *Fula) Load(ctx context.Context, l ipld.Link, np ipld.NodePrototype) (ipld.Node, error) {
+	return f.ls.Load(ipld.LinkContext{Ctx: ctx}, l, np)
+}
+
+func (f *Fula) Has(ctx context.Context, l ipld.Link) (bool, error) {
+	return f.ds.Has(ctx, toDatastoreKey(l))
+}
+
+func toDatastoreKey(l ipld.Link) datastore.Key {
+	return datastore.NewKey(l.Binary())
+}
+
+func (f *Fula) Fetch(ctx context.Context, from peer.ID, l ipld.Link) error {
+	if exists, err := f.Has(ctx, l); err != nil {
 		return err
+	} else if exists {
+		return nil
 	}
+	resps, errs := f.gx.Request(ctx, from, l, selectorparse.CommonSelector_ExploreAllRecursively)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case resp, ok := <-resps:
+			if !ok {
+				return nil
+			}
+			log.Infow("synced node", "node", resp.Node)
+		case err, ok := <-errs:
+			if !ok {
+				return nil
+			}
+			log.Warnw("sync failed", "err", err)
```

Comment:
We probably should revise error handling here, considering function signature returns error. but this is a good place to start.